lib/datadog/statsd/sender.rb

```diff
@@ -32,40 +46,56 @@ def rendez_vous
       def add(message)
         raise ArgumentError, 'Start sender first' unless message_queue
 
+        # if the thread does not exist, we assume we are running in a forked process,
+        # empty the message queue and message buffers (these messages belong to
+        # the parent process) and spawn a new companion thread.
+        if !sender_thread.alive?
+          @logger.warn { "Statsd: companion thread dead, re-creating one" } if @logger
+          @mx.synchronize {
+            message_queue.close if CLOSEABLE_QUEUES
+            @message_queue = nil
+            message_buffer.reset
+            start
```

Comment:
There's a race hiding here. If a process just forked, and started two new threads, and both threads try to report data, the following timeline can occur:

1. Thread A checks line 52, sees that no sender thread is alive. Grabs the lock. (Scheduler switches threads)
2. Thread B checks line 52, sees that no sender thread is alive. Tries to grab the lock, is taken, so it blocks waiting for lock (Scheduler switches threads)
3. Thread A closes previous queue and resets it. Creates a new thread. Releases the lock. Adds message to queue. (Scheduler switches threads)
4. Thread B wakes up, grabs the lock, closes the queue that Thread A created and resets it (data from Thread A) gets lost. Creates another sender thread. etc...

TL;DR After grabbing the lock, we need to check that the world state is as we expected it to be when we went in to grab the lock, as it might've moved in the meanwhile :)

Reply:
> If a process just forked, and started two new threads

I don't expect the first process to have to create a thread, but I agree it could happen if for an unknown reason the first process see its thread dying at the same moment. I'll look into this.